pygmt/helpers/utils.py

```diff
@@ -61,23 +66,41 @@ def data_kind(data, x=None, y=None, z=None, required_z=False):
     'file'
     >>> data_kind(data=pathlib.Path("my-data-file.txt"), x=None, y=None)
     'file'
+    >>> data_kind(data=None, x=None, y=None, optional_data=True)
+    'null'
+    >>> data_kind(data=2.0, x=None, y=None, optional_data=True)
+    'null'
+    >>> data_kind(data=True, x=None, y=None, optional_data=True)
+    'null'
     >>> data_kind(data=xr.DataArray(np.random.rand(4, 3)))
     'grid'
     """
-    if data is None and x is None and y is None:
-        raise GMTInvalidInput("No input data provided.")
-    if data is not None and (x is not None or y is not None or z is not None):
-        raise GMTInvalidInput("Too much data. Use either data or x and y.")
-    if data is None and (x is None or y is None):
+    # pylint: disable=too-many-branches
+    # validate the combinations of data/x/y/z
+    if x is None and y is None:  # both x and y are not given
+        if data is None and not optional_data:
+            raise GMTInvalidInput("No input data provides")
+        if data is not None and z is not None:
+            raise GMTInvalidInput("Too much data. Use either data or x/y/z.")
+    elif x is None or y is None:  # either x or y is not given
         raise GMTInvalidInput("Must provide both x and y.")
-    if data is None and required_z and z is None:
-        raise GMTInvalidInput("Must provide x, y, and z.")
+    else:  # both x and y are given
+        if data is not None:
+            raise GMTInvalidInput("Too much data. Use either data or x and y.")
+        if required_z and z is None:
+            raise GMTInvalidInput("Must provide x, y, and z.")
 
+    # determine the data kind
     if isinstance(data, (str, pathlib.PurePath)):
         kind = "file"
+    elif optional_data and (data is None or isinstance(data, (bool, int, float))):
+        # a nullcontext will be created for "null" kind
+        kind = "null"
```

Comment:
So for `shading="+a30+nt0.8"` (a str type argument), the returned kind will be `file` instead of `null`. Technically both `file` and `null` returns a nullcontext from `virtualfile_from_data` so it shouldn't really matter, but it makes me wonder if we should combine the `file` and `null` kind, or be a bit more explicit somehow, otherwise this `data_kind` function will get confusing to new maintainers.

Reply:
> So for `shading="+a30+nt0.8"` (a str type argument), the returned kind will be `file` instead of `null`.

Yes, maybe we can rename `file` to `file_or_arg`?

Reply:
Naming is hard :upside_down_face: Yeah, let's go with `file_or_arg`.

Reply:
Done in 9a43dc8edab2b5b0f907b84e0ff740cda2c4c80e.

Reply:
> So for `shading="+a30+nt0.8"` (a str type argument), the returned kind will be `file` instead of `null`.

Technically speaking, `"+a30+nt0.8"` is a valid file name, so a `file` kind still makes sense. 

The `data_kind` function is also being used in many module wrappers. For example, in `legend`,

https://github.com/GenericMappingTools/pygmt/blob/6764554c84e9b4c328b3b41d60c8337fa0f18373/pygmt/src/legend.py#L78-L83

`specfile` can be either `None` or a "file" (string type), but can't be in bool/int/float type. Thus, instead of having a single `file_or_arg` kind, we may need to have more kinds like `file` (str or PurePath), `null` (None), `bool` (True or False), `value` (int/float).

Reply:
Ok, maybe just keeping `"file"` kind is good rather than `"file_or_arg"`, and we can add the `"null"` kind back? Not sure if adding separate kinds for `bool` or `value` is needed, since all of these would just return a nullcontext.

Reply:
Have two different kinds `file` and `arg` in commit 5faa05dc27b458e7ba675e02704faefcf05d9086